File: src/public/RestorationResult.ts
export abstract class RestorationResult {
  abstract toJson(): Object;

  static restored() {
    return new Restored();
  }

  static failed(error?: Error) {
    return new Failed(error);
  }
}

export class Restored extends RestorationResult {
  toJson() {
    return { result: 'restored' };
  }
}

export class Failed extends RestorationResult {
  constructor(public error?: Error) {
    super();
  }

  toJson() {
    return {
      result: 'failed',
      errorMessage: this.error ? this.error.message : null,
    };
  }
}

File: android/src/main/java/com/superwallreactnative/models/RestorationResult.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.delegate.RestorationResult

class RestorationResult {
  companion object {
    fun fromJson(json: ReadableMap): RestorationResult {
      return when (json.getString("result")) {
        "restored" -> RestorationResult.Restored()
        "failed" -> {
          val errorMessage = json.getString("errorMessage") ?: "Unknown error"
          RestorationResult.Failed(Error(errorMessage))
        }
        else -> RestorationResult.Failed(Error("Unknown restoration result"))
      }
    }
  }
}
